lantern_embeddings/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+use csv::Writer;
+use lantern_embeddings_core::clip;
+use rand::Rng;
+use std::io::Write;
+use std::sync::mpsc;
+use std::sync::mpsc::{Receiver, Sender};
+use std::thread::JoinHandle;
+use std::time::Instant;
+use std::{panic, process};
+
+use postgres::{Client, NoTls, Row};
+
+pub mod cli;
+
+type EmbeddingRecord = (String, Vec<f32>);
+
+fn producer_worker(
+    args: &cli::EmbeddingArgs,
+    tx: Sender<Vec<Row>>,
+) -> Result<JoinHandle<()>, anyhow::Error> {
+    let uri = args.uri.clone();
+    let pk = args.pk.clone();
+    let column = args.column.clone();
+    let schema = args.schema.clone();
+    let table = args.table.clone();
+    let batch_size = args.batch_size.clone();
+
+    let handle = std::thread::spawn(move || {
+        let mut client = Client::connect(&uri, NoTls).unwrap();
+        let mut transaction = client.transaction().unwrap();
+        let rows = transaction
+            .query(
+                &format!(
+                    "SELECT reltuples::bigint AS estimate FROM pg_class WHERE oid ='{}'::regclass",
+                    table
+                ),
+                &[],
+            )
+            .unwrap();
+        let count: i64 = rows[0].get(0);
+        println!(
+            "[*] Found approximately {} items in table \"{}\"",
+            count, table
+        );
+        // With portal we can execute a query and poll values from it in chunks
+        let portal = transaction
+            .bind(
+                &format!("SELECT \"{pk}\"::text, \"{column}\" FROM \"{schema}\".\"{table}\";"),
+                &[],
+            )
+            .unwrap();
+
+        loop {
+            // poll batch_size rows from portal and send it to embedding thread via channel
+            let rows = transaction
+                .query_portal(&portal, batch_size as i32)
+                .unwrap();
+
+            if rows.len() == 0 {
+                break;
+            }
+
+            if tx.send(rows).is_err() {
+                break;
+            }
+        }
+        drop(tx);
+    });
+
+    return Ok(handle);
+}
+
+fn embedding_worker(
+    args: &cli::EmbeddingArgs,
+    rx: Receiver<Vec<Row>>,
+    tx: Sender<Vec<EmbeddingRecord>>,
+) -> Result<JoinHandle<()>, anyhow::Error> {
+    let is_visual = args.visual.clone();
+    let model = args.model.clone();
+    let data_path = args.data_path.clone();
+    let mut count: u64 = 0;
+
+    let handle = std::thread::spawn(move || {
+        let mut start = Instant::now();
+        loop {
+            let rows = rx.recv();
+            if rows.is_err() {
+                // channel has been closed
+                break;
+            }
+
+            if count == 0 {
+                // mark exact start time
+                start = Instant::now();
+            }
+
+            let rows = rows.unwrap();
+            let mut input_vectors = Vec::with_capacity(rows.len());
+
+            for row in &rows {
+                let col: &str = row.get(1);
```

Comment:
expect some input text fields to be empty. This probably should be read as Option<&str>.
It is probably fine to populate the corresponding embedding field with the embedding corresponding to the empty string. 
I suspect this is the simplest approach, otherwise, you'd need to change the vector of strings to vector of (primaryKey, string) pairs.

Reply:
for now added one more vector for ids, I will need to refactor this later to support failure of a single image download/read while generating image embeddings.